django/db/models/fields/json.py

```diff
@@ -70,8 +70,11 @@ def deconstruct(self):
     def from_db_value(self, value, expression, connection):
         if value is None:
             return value
-        if connection.features.has_native_json_field and self.decoder is None:
-            return value
+        if connection.features.has_native_json_field:
+            if self.decoder is None:
+                return value
+            else:
+                value = json.dumps(value, cls=self.encoder)
```

Comment:
So this will mean a performance penalty on such databases:

1. psycopg2 decodes the data
2. here we encode the data
3. line 79 decodes the data with the encoder

Is there no way around this? Perhaps rather than a `::text` cast, it's possible to use a different approach to select it as text, or prevent psycopg2 from parsing the JSON?

Reply:
According to `psycopg2` docs we could register dummy `loads()` function :
```python
psycopg2.extras.register_default_json(loads=lambda x: x)
```

Reply:
We'd actually need to use [`psycopg2.extras.register_default_jsonb()`](https://www.psycopg.org/docs/extras.html#psycopg2.extras.register_default_jsonb). Also, unless you pass a connection or cursor, you need to pass `globally=True`. Someone might be relying on the default behaviour, so we should be explicit and set it on a connection. I guess that could be done in a `connection_created` signal?

My only concern is that this might break some applications that touch the raw underlying connections. It does feel like a blunt hammer.

Ideally we'd somehow just wrap the section where the value is adapted with calls to this function to set/reset.

Reply:
> My only concern is that this might break some applications that touch the raw underlying connections. It does feel like a blunt hammer.

Yes this is a bit backwards incompatible for apps using raw SQL using Djanogo's connection to select from a jsonb column. I'm not sure it's worth the effort though to try wrap specific queries though, that would require a new field -> queryset interface.

Reply:
> for apps using raw SQL

No need to use raw SQL.  I was lazy and used `JSONField` (doesn't require a `base_field`) instead of `ArrayField` as `output_field` in the following query which resulted in a TypeError (list instead of str/bytes/bytearray passed to `json.loads()`).

```python
class Project(Model):
    project_id = CharField(unique=True, max_length=36)


class Lexicon(Model):
    lemma = CharField(max_length=256, db_index=True)
    # lexicon entries can be global or specific to a set of projects
    projects = ManyToManyField(Project, through="LexiconProjectThrough")


class LexiconProjectThrough(Model):
    lexicon = ForeignKey(AdjectiveLexicon, on_delete=CASCADE)
    project = ForeignKey(Project, on_delete=CASCADE)

    class Meta:
        unique_together = (("lexicon", "project"),)


Lexicon.objects.annotate(
    project_specific=Exists(
        LexiconProjectThrough.objects.filter(lexicon=OuterRef("pk"))
    )
).annotate(
    project_ids=Case(
        When(project_specific=True, then=ArrayAgg("projects__project_id")),
        default=[],
        output_field=JSONField(),
    )
).first()
```

The solution was to remove `output_field=JSONField()` (or replace it with `output_field=ArrayField(base_field=CharField(max_length=36))`).  Back in 2017 (Django 1.11) the code failed without the output_field (`TypeError: int() argument must be a string, a bytes-like object or a number, not 'list'`; project_id was a PositiveIntegerField back then).

The code also works when using a custom `JSONField` that brings back the `self.decoder is None` check:

```python
class MyJSONField(JSONField):
    def from_db_value(self, value, expression, connection):
        if connection.features.has_native_json_field and self.decoder is None:
            return value
        return super().from_db_value(value, expression, connection)
```

<details><summary>stack trace</summary>

```
/usr/local/lib/python3.8/site-packages/django/db/models/query.py in first(self)
    676     def first(self):
    677         """Return the first object of a query or None if no match is found."""
--> 678         for obj in (self if self.ordered else self.order_by('pk'))[:1]:
    679             return obj
    680

/usr/local/lib/python3.8/site-packages/django/db/models/query.py in __iter__(self)
    285                - Responsible for turning the rows into model objects.
    286         """
--> 287         self._fetch_all()
    288         return iter(self._result_cache)
    289

/usr/local/lib/python3.8/site-packages/django/db/models/query.py in _fetch_all(self)                                                         [27/2653]
   1301     def _fetch_all(self):
   1302         if self._result_cache is None:
-> 1303             self._result_cache = list(self._iterable_class(self))
   1304         if self._prefetch_related_lookups and not self._prefetch_done:
   1305             self._prefetch_related_objects()

/usr/local/lib/python3.8/site-packages/django/db/models/query.py in __iter__(self)
     68             ])) for field, related_objs in queryset._known_related_objects.items()
     69         ]
---> 70         for row in compiler.results_iter(results):
     71             obj = model_cls.from_db(db, init_list, row[model_fields_start:model_fields_end])
     72             for rel_populator in related_populators:

/usr/local/lib/python3.8/site-packages/django/db/models/sql/compiler.py in apply_converters(self, rows, converters)
   1098                 value = row[pos]
   1099                 for converter in convs:
-> 1100                     value = converter(value, expression, connection)
   1101                 row[pos] = value
   1102             yield row

/usr/local/lib/python3.8/site-packages/django/db/models/fields/json.py in from_db_value(self, value, expression, connection)
     72             return value
     73         try:
---> 74             return json.loads(value, cls=self.decoder)
     75         except json.JSONDecodeError:
     76             return value

/usr/local/lib/python3.8/json/__init__.py in loads(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    339     else:
    340         if not isinstance(s, (bytes, bytearray)):
--> 341             raise TypeError(f'the JSON object must be str, bytes or bytearray, '
    342                             f'not {s.__class__.__name__}')
    343         s = s.decode(detect_encoding(s), 'surrogatepass')

TypeError: the JSON object must be str, bytes or bytearray, not list
```

</details>

----

That said, at the moment I don't think that's worth an issue, but I wanted to let you know in case you think otherwise. :)

Reply:
In the first case I think you were just lucky that the JSONField effectively was "pass through" for your return data.

In the second case - I don't really think it's concern since `decoder` was introduced in Django 3.1, and users who subclass and duplicate logic become responsible for their duplicated logic staying in sync.

Reply:
@michael-k As far as I'm aware you should use `JSONBAgg()` instead of `ArrayAgg()` if you want to return a ``JSON`` array.